autotest/test_prt_libmf6_budget.py

```diff
@@ -0,0 +1,110 @@
+"""
+This test runs the simulations in test_prt_budget.py, but uses the api
+to run the PRT test.
+"""
+
+from pathlib import Path
+
+import pytest
+from framework import TestFramework
+from modflowapi import ModflowApi
+from test_prt_budget import (
+    HorizontalCase,
+    build_mp7_sim,
+    build_prt_sim,
+    check_output,
+)
+
+simname = "prt_libmf6"
+cases = [simname]
+
+
+def build_models(idx, test):
+    # build MODFLOW 6 files
+    ws = test.workspace
+    name = cases[idx]
+
+    gwf_sim = HorizontalCase.get_gwf_sim(
+        test.name, test.workspace, test.targets["mf6"]
+    )
+    prt_sim = build_prt_sim(
+        test.name,
+        test.workspace,
+        test.workspace / "prt",
+        test.targets["libmf6"],
+    )
+    mp7_sim = build_mp7_sim(
+        test.name,
+        test.workspace / "mp7",
+        test.targets["mp7"],
+        gwf_sim.get_model(),
+    )
+
+    return gwf_sim, prt_sim, mp7_sim
+
+
+def api_func(exe, idx, model_ws=None):
+    name = cases[idx].upper()
+    if model_ws is None:
+        model_ws = Path(".")
+
+    output_file_path = model_ws / "mfsim.stdout"
+
+    try:
+        mf6 = ModflowApi(exe, working_directory=model_ws)
+    except Exception as e:
+        print("Failed to load " + exe)
+        print("with message: " + str(e))
+        return False, open(output_file_path).readlines()
+
+    # initialize the model
+    try:
+        mf6.initialize()
+    except:
+        return False, open(output_file_path).readlines()
+
+    # time loop
+    current_time = mf6.get_current_time()
+    end_time = mf6.get_end_time()
+
+    # model time loop
+    idx = 0
+    while current_time < end_time:
+        # get dt and prepare for non-linear iterations
+        dt = mf6.get_time_step()
+        mf6.prepare_time_step(dt)
+
+        # convergence loop
+        kiter = 0
+        mf6.prepare_solve()
```

Comment:
The reason to have prepare_solve, solve, finalize_solve is so you can call solve multiple times and manipulate in between. Is that something you would want to support with this Explicit Solution? And if so, should it maybe also be a test? In this particular test I guess you could change everything with a call to update.

Reply:
Very good point.  This is a work in progress, and we are not quite ready to be able to call solve multiple times yet for PRT.  Your comment did also point out that PRT won't work quite right with ATS yet (particles wouldn't get reset if a time step needs to be retried with a shorter length and output is written during the solve).  This has triggered  discussions for how to support PRT/ATS and PRT with multiple solves.  For now, this new support for the explicit solution makes it possible for modflowapi to run any of the PRT models (without any manipulation between the different solve steps).  The current version of modflowapi run_model always uses the lowest level XMI calls (preparesolve, solve, finalizesolve) instead of preparetimestep, dotimestep, finalizetimestep or or update.